### ISSUE TITLE
Support for python lambda functions on AWS

### ISSUE THREAD

**Issue by kdenz** [sst/ion#366]
Hi! Love SST, thank you so much for all the work over the years! Curious when python lambda functions will be supported for AWS?

**Comment by thdxr** [comment]
our challenge is no one on the team uses python day to day so it's quite hard for us to create a good python experience given how many different ways there is to do things in that ecosystem<br><br>nothing stops us from supporting it technically we just don't have much clarity on what supporting it well means

**Comment by RamblinGambler** [comment]
Is it possible to document a working python function pattern using the pulumi api in the ion run block? That way the dev team doesn't have to commit  to building an experience they are unfamiliar with and on the flipside this would do two things:

1. Give people who just want to run some python a pattern to follow. It doesn't have to be the best experience like the rest of the native ion behavior but its something for devs to go off of and extend.

2. The more people use it the more interest there will be to improve it.

I am having a bear of a time getting python pulumi functions to be recognized in the ion run block. I could use some better guidance than the scattered bits of info between the sites and discord, but once it works I can help improve it.

**Comment by cmanou** [comment]
The way i've been working around it from a purely deployment point of view with ion, which still has problems with dev experience is via docker image lambdas. I have a docker image based of the python lambda base images (since I found its easiest way to deal with dependencies for platform specific) and then use code similar to this in the run block

```
const imageRepository = new aws.ecr.Repository(
      "ImageRepository",
      { name: "<name_repository>" }
);

const imageAuthToken = aws.ecr.getAuthorizationTokenOutput({
  registryId: imageRepository.registryId,
});
const image = new docker.Image("Image", {
  imageName: $interpolate`${imageRepository.repositoryUrl}:latest`,
  build: {
    context: path.join($cli.paths.root, "projects/<path_to_docker_context>"),
    platform: "linux/amd64",
  },
  registry: {
    server: imageAuthToken.proxyEndpoint,
    username: imageAuthToken.userName,
    password: $util.secret(imageAuthToken.password),
  },
});
const lambdaRole = new aws.iam.Role("LambdaRole", {
  assumeRolePolicy: aws.iam.assumeRolePolicyForPrincipal({
    Service: "lambda.amazonaws.com",
  }),
  inlinePolicies: [],
  managedPolicyArns: [
    "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
  ],
});

const lambdaHandler = new aws.lambda.Function(`HandlerFunction`, {
  name: `${$app.name}-${$app.stage}-HandlerFunction`,
  imageUri: image.repoDigest,
  timeout: 120,
  architectures: ["x86_64"],
  packageType: "Image",
  role: lambdaRole.arn,
});
```
But yeah better support from a dev experience side would be great. Though I agree with @thdxr that the ecosystem has many varied options for things so supporting all of them will be an up hill battle and there doesn't seem to be any leading contenders especially in the dependency management/bundling side of things.

Potentially a middle ground could be something like first class docker image support for SST functions and live mode could maybe do something with running the images in the local test mode that spins up a http server but that would only work if you use the proper base images on the runtime interface client.

**Comment by kdenz** [comment]
Much agreed! I would look forward to just feature parity with SST v2's Function, where it at least allows easy development and deployment of python lambda functions, with live debugging / breakpoint debugging, etc. Just wondering if there's any timeline for SST Ion's Function to support python runtimes like SST v2? @thdxr

**Comment by partmor** [comment]
in theory it should be possible to have **live** python functions as of now with `provided.al2023` and custom `bundler` and `handler` (like in the Swift example), right?

(cc @thdxr )

**Comment by subssn21** [comment]
I use v2 for both of my production products with Python Lambdas without any issues, so a great first step would be feature parity. If there’s any testing or anything that I can help with, more than happy to help out.

if we are talking new features, the only thing I would like to see is the packages installed by running the pip install requirements command be packaged in a layer so they only have to be built once instead of for each function, but that’s a nice to have

**Comment by jayair** [comment]
Somebody's got a PR in the works https://github.com/sst/ion/pull/931

**Comment by subwaymatch** [comment]
Python Lambda (+dockerized) support would be a killer feature for Ion! I asked the same question a few weeks ago on Discord and learned that this is in the roadmap, although not urgent. This was the only missing feature that left me with CDK at the moment. 

![image](https://github.com/user-attachments/assets/44991a49-391f-42dd-81ed-b933afa4f24b)
![image](https://github.com/user-attachments/assets/dd9673af-8ce6-4ffb-a016-6c28ff79569d)

**Comment by jayair** [comment]
The PR got merged. We just need to support Resource Links and we can launch this.

**Comment by subssn21** [comment]
Any way we can add a parameter to change the PythonPath, It looks like right now, it is all relative to the handler function and my v2 app (and I am sure others) have their library code in a different directory than their handlers. 
As an example
Project Root
|-backend
  |-lib
     |-models
     |-utils
  |-services
    |-api
      -handler1
      -handler2
   |-dynamodbstream
     -handler1
     -handler2
  |-opensearch  
|-frontend
-sst.config.ts

Does that make sense?

**Comment by jakevollkommer** [comment]
> Somebody's got a PR in the works #931

[migrate from v2 docs](https://sst.dev/docs/migrate-from-v2/) still say non-node runtimes are still not supported with `Function`

but sounds like python support was merged 3 weeks ago, wondering what we should still be waiting on?

we want to migrate to v3 but not having python runtimes is a huge blocker because a significant portion of our backend is in python

**Comment by jayair** [comment]
Yeah there's still a couple of things we need to fix with it. But it's almost ready.

**Comment by jayair** [comment]
But you can give it a try.

**Comment by aeftimia** [comment]
Just noting the example [aws-python](https://github.com/sst/ion/tree/dev/examples/aws-python) is not reproducible. I get an import error on `import Resource from sst` with and without the line `sst = { git = "https://github.com/walln/ion.git", subdirectory = "sdk/python", branch = "walln/python-fixes" }` in `pyproject.toml`. Also it seems like no matter what the configuration is within `pyproject.toml` and the lambda runtime in `sst.config.ts`, the system always installs and runs python 3.13. I've seen this behavior in my own codebase as well. 

You can easily verify this by modifying the aws-python example's python.py file with the following code

```
import sys
# from sst import Resource


def handler(event, context):
    print("Function invoked from Python", sys.version)

    return {
        "statusCode": 200,
        # "body": f"{Resource.MyLinkableValue.foo} from Python!",
    }
```
